tests/unit/test_schedule.cpp

```diff
@@ -0,0 +1,280 @@
+#include <algorithm>
+#include <random>
+#include <stdexcept>
+#include <vector>
+
+#include <common_types.hpp>
+#include <schedule.hpp>
+#include <util/partition.hpp>
+#include <util/rangeutil.hpp>
+
+#include "common.hpp"
+#include "stats.hpp"
+
+using namespace nest::mc;
+using namespace testing;
+
+// Pull events from n non-contiguous subintervals of [t0, t1)
+// and check for monotonicity and boundedness.
+
+void run_invariant_checks(schedule S, time_type t0, time_type t1, unsigned n, int seed=0) {
+    if (!n) return;
+
+    std::minstd_rand R(seed);
+    std::uniform_real_distribution<time_type> U(t0, t1);
+
+    std::vector<time_type> divisions = {t0, t1};
+    std::generate_n(std::back_inserter(divisions), 2*(n-1), [&] { return U(R); });
+    util::sort(divisions);
+
+    bool skip = false;
+    for (auto ival: util::partition_view(divisions)) {
+        if (!skip) {
+            auto ts = S.events(ival.first, ival.second);
+
+            EXPECT_TRUE(std::is_sorted(ts.begin(), ts.end()));
+            if (!ts.empty()) {
+                EXPECT_LE(t0, ts.front());
+                EXPECT_GT(t1, ts.back());
+            }
+        }
+        skip = !skip;
+    }
+}
+
+// Take events from n contiguous intervals comprising [t0, t1), reset, and
+// then compare with events taken from a different set of contiguous
+// intervals comprising [t0, t1).
+
+void run_reset_check(schedule S, time_type t0, time_type t1, unsigned n, int seed=0) {
+    if (!n) return;
+
+    std::minstd_rand R(seed);
+    std::uniform_real_distribution<time_type> U(t0, t1);
+
+    std::vector<time_type> first_div = {t0, t1};
+    std::generate_n(std::back_inserter(first_div), n-1, [&] { return U(R); });
+    util::sort(first_div);
+
+    std::vector<time_type> second_div = {t0, t1};
+    std::generate_n(std::back_inserter(second_div), n-1, [&] { return U(R); });
+    util::sort(second_div);
+
+    std::vector<time_type> first;
+    for (auto ival: util::partition_view(first_div)) {
+        util::append(first, S.events(ival.first, ival.second));
+    }
+
+    S.reset();
+    std::vector<time_type> second;
+    for (auto ival: util::partition_view(second_div)) {
+        util::append(second, S.events(ival.first, ival.second));
+    }
+
+    EXPECT_EQ(first, second);
+}
+
+TEST(schedule, regular) {
+    // Use exact fp representations for strict equality testing.
+    std::vector<time_type> expected = {0, 0.25, 0.5, 0.75, 1.0};
+
+    schedule S = regular_schedule(0.25);
+    EXPECT_EQ(expected, S.events(0, 1.25));
+
+    S.reset();
+    EXPECT_EQ(expected, S.events(0, 1.25));
+
+    S.reset();
+    expected = {0.25, 0.5, 0.75, 1.0};
+    EXPECT_EQ(expected, S.events(0.1, 1.01));
+}
+
+TEST(schedule, regular_invariants) {
+    SCOPED_TRACE("regular");
```

Comment:
I would use `SCOPED_TRACE("regular_invariants")` etc. to make hunting through the tests easier?

Reply:
Good call.